src/OrchardCore/OrchardCore.Users.Core/Services/DefaultUserClaimsPrincipalFactory.cs

```diff
@@ -12,20 +14,33 @@ namespace OrchardCore.Users.Services
     /// </summary>
     public class DefaultUserClaimsPrincipalFactory : UserClaimsPrincipalFactory<IUser, IRole>
     {
+
+        private readonly IServiceProvider _serviceProvider;
+
         public DefaultUserClaimsPrincipalFactory(
             UserManager<IUser> userManager,
             RoleManager<IRole> roleManager,
-            IOptions<IdentityOptions> identityOptions) : base(userManager, roleManager, identityOptions)
+            IOptions<IdentityOptions> identityOptions,
+            IServiceProvider serviceProvider) : base(userManager, roleManager, identityOptions)
```

Comment:
Just inject `IEnumerable<IUserClaimsProvider>` here instead of the service provider